### PR TITLE
InputSystem fixes and features

### LINKED ISSUE #966: Opening the interaction wheel does not prevent from interacting with the rest of the world.
## Summary

Currently, when opening the interaction wheel, the player can still interact with other stuff, like picking up items, honking ... 
It would make more sense to me if the interaction wheel would prevent anything but choosing an interaction from it.

## Expected Behavior

When the interaction wheel is open, the player should not be able to do anything else but choose an interaction from it.
The player should not be able to pickup or drop items by left clicking on them in the world.

### LINKED ISSUE #1003: Make the right click for interactions great again
## Summary

Currently, holding right click over an interaction on the interaction wheel and releasing the right click does nothing.
It would be great, if simply releasing the right click would activate the interaction, instead of having to left click.

image

Everything you need to change will probably be in radialInteractionButton.cs

### LINKED ISSUE #1073: drag and drop item sprite does not work anymore
## Summary

Right now when clicking on a sprite in the UI, when holding an object, the item just fall on the floor. It looks like the UI is ignored. 

## Steps to Reproduce

- Pick up an item
- Try to drag and drop the sprite on the other hand.
- Fail
- ....
- Profit

## Possible Solution

This had something to do with the new input system, looking at that commit could help "Replace Input System with Input Manager"

### LINKED ISSUE #1090: Tilemap construction doesn't prevent non-tilemap interaction PART 2
## Summary

The tilemap construction window doesn't prevent interaction with the rest of the world. Part 2.

## Current Behavior

<kbd>Shift</kbd> is used to toggle walking/running. It is also used by the tilemap editor for replacing tiles (hold <kbd>Shift</kbd> and click <kbd>LMB</kbd>).

The issue is, that when you are building a map and using the replace feature, you end up also changing your walk/run toggle repeatedly which becomes annoying while building a map.

## Steps to Reproduce

- Open the tilemap window and select a floor tile
- Move back and forth to confirm your current walk/run state
- Hold <kbd>Shift</kbd> and click <kbd>LMB</kbd> to replace a few floor tiles
- Move back and forth to confirm your walk/run state has changed

## Expected Behavior

Non-tilemap interactions should be ignored in favor of tilemap interactions while the tilemap editor is open.

## Context

Related issues
#966 
#1086
#1091 

## Media

https://user-images.githubusercontent.com/41941008/223557693-166ae765-47c3-4b88-8c73-1c509952565b.mp4

### LINKED ISSUE #1091: Tilemap construction doesn't prevent non-tilemap interaction PART 3
## Summary

The tilemap construction window doesn't prevent interaction with the rest of the world. Part 3.

## Current Behavior

You can pick up and drop items while using the tilemap editor.

## Steps to Reproduce

- Open tilemap editor and select a floor tile under the plenum layer,
- Try replace a floor tile (hold <kbd>Shift</kbf> and click <kbd>LMB</kbd>) beneath an item on the ground (actually click the item),
- Notice you picked up the item while also replacing the floor tile,
- Replace another floor tile nearby,
- Notice you place the item down again while also replacing the floor tile.

## Expected Behavior

Non-tile map interactions should be ignored in favor of tilemap interactions while the tilemap editor is open.

## Context

Related issues
#966 
#1086
#1090 

## Media

https://user-images.githubusercontent.com/41941008/223558929-4f939037-da27-40d2-b2b3-a0a2cf56a343.mp4

### PR DESCRIPTION
**Pull request by iamteapot422**
## Summary
Adds toggling methods to InputSystem to prevent any issues with this.
Replaces all Disable() and Enable() with ToggleMethods
Moves InputSystem to SS3D.Systems.Inputs namespace
Moves InputSystem game object from Lobby.scene to Startup.scene and adds DontDeleteOnLoad to it, removes InputSystem.cs from Camera in Startup.scene. This is necessary, because without these changes, SystemLocator registers only InputSystem.cs in Startup.scene, thats gets deleted in Lobby.cs.
Also enables Input Consumption in Input Action asset. This feature prevents overlap of actions with modifiers. For example, if you have CTRL+E and E actions, if you press CTRL firstly and E button secondly, E action won't be triggered

## Fixes 
Closes #966 
Closes #1073 
Closes #1090 
Closes #1091 
Closes #1003

### CODE REVIEW

File: Assets/Scripts/SS3D/Systems/Entities/Silicon/ThreadController.cs [hunk 1]
```diff
@@ -130,9 +133,13 @@ private void MovePlayer()
         /// <returns></returns>
         private void ProcessPlayerInput()
         {
-            float x = _controls.Movement.ReadValue<Vector2>().x;
-            float y = _controls.Movement.ReadValue<Vector2>().y;
-
+            float x = 0;
+            float y = 0;
+            if (_controls.Movement.phase == InputActionPhase.Performed)
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Systems/Entities/Silicon/ThreadController.cs:140]: can you explain what's "phase" and InputActionPhase ?

**Review by stilnat — COMMENTED** [review]
I did a partial review, I left the big changes on the side for now. Going to play test that a bit.

File: Assets/Scripts/SS3D/Systems/Entities/Silicon/ThreadController.cs (hunk 1 above)
Comment (iamteapot422) [reply Assets/Scripts/SS3D/Systems/Entities/Silicon/ThreadController.cs:140]: Every action switches between five phases: Disabled, Waiting, Started, Perfromed, Canceled. For example, if you have hold(0.5 seconds) interaction on an action, the action has Performed phase when you hold a key for 0.5 seconds or more

File: Assets/Scripts/SS3D/Systems/Interactions/InteractionController.cs [hunk 2]
```diff
@@ -38,32 +41,58 @@ public override void OnStartClient()
...
         }
 
+        /// <summary>
+        /// Runs the most prioritised action
+        /// </summary>
         [Client]
-        private void HandlePrimaryClick(InputAction.CallbackContext callbackContext)
+        private void HandlePerform(InputAction.CallbackContext callbackContext)
         {
-            RunPrimaryInteraction();
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Systems/Interactions/InteractionController.cs:61]: I like the RunPrimaryInteraction, we just have to read one line to understand what's happening and it's just more modular if we want to modify it later on. Could you add it back ?

Comment (stilnat) [inline Assets/Scripts/SS3D/Systems/Interactions/InteractionController.cs:61]: I like the RunPrimaryInteraction, we just have to read one line to understand what's happening and it's just more modular if we want to modify it later on. Could you add it back ?

File: Assets/Scripts/SS3D/Systems/Interactions/InteractionController.cs [hunk 3]
```diff
@@ -38,32 +41,58 @@ public override void OnStartClient()
...
+            if (EventSystem.current.IsPointerOverGameObject())
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Systems/Interactions/InteractionController.cs:72]: You can leave the check in HandlePerform though

File: Assets/Scripts/SS3D/Systems/Interactions/InteractionController.cs [hunk 4]
```diff
@@ -38,32 +41,58 @@ public override void OnStartClient()
...
-            _otherControls.PrimaryClick.performed -= HandlePrimaryClick;
-            _otherControls.SecondaryClick.performed -= HandleSecondaryClick;
+            _controls.PerformInteraction.performed -= HandlePerform;
+            _controls.ViewInteractions.performed -= HandleView;
             _hotkeysControls.Use.performed -= HandleUse;
+            _inputSystem.ToggleActionMap(_controls, false);
         }
 
+        /// <summary>
+        /// Runs the most prioritised action
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Systems/Interactions/InteractionController.cs]: replace action with interaction.

**Review by stilnat — CHANGES_REQUESTED** [review]
another partial review, three more files to go

File: Assets/Scripts/SS3D/Systems/Interactions/InteractionController.cs (hunk 2 above)
Comment (iamteapot422) [reply Assets/Scripts/SS3D/Systems/Interactions/InteractionController.cs:61]: I'll better rename action to Run Primary

**Comment by stilnat** [comment]
@Mechar418 I got a bug, weirdly enough, only when posing the blue soda cans in tile menu on the floor . It temporarily freeze the game. Can you check if this happens on develop ?  
image

**Comment by stilnat** [comment]
@Mechar418 can't run anymore with tile menu opened. Would be nice to keep that one.

**Comment by iamteapot422** [comment]
@stilnat sure

**Comment by iamteapot422** [comment]
@stilnat you can't run, because shift is already binded to "replace" in TileMapCreator. If you close the window, you will be able to run again

**Comment by stilnat** [comment]
> @stilnat you can't run, because shift is already binded to "replace" in TileMapCreator. If you close the window, you will be able to run again

right, but it used to work before ? can't we bind it to both toggle run and replace ?

**Comment by stilnat** [comment]
@Mechar418 I got this bug, is that your known issue ?

> NullReferenceException: Object reference not set to an instance of an object
> SS3D.Systems.Tile.TileMapCreator.GhostManager.ChangeGhostColor (SS3D.Systems.Tile.TileMapCreator.GhostManager+BuildMatMode mode) (at Assets/Scripts/SS3D/Systems/Tile/TileMapCreator/GhostManager.cs:92)
> SS3D.Systems.Tile.UI.TileMapCreator.RpcLogic___RpcReceiveCanBuild_214505783 (FishNet.Connection.NetworkConnection conn, System.Boolean canBuild) (at Assets/Scripts/SS3D/Systems/Tile/TileMapCreator/TileMapCreator.cs:247)
> 

I got it after playing around with the build menu, then grabbing stuff and posing them again on the floor. 
It's hard to reproduce this one, you have to mess around quite a bunch for it to happen.
After I get the exception, I can't put the item I have in hand on the floor, and everytime I try I get the same exception.
I reproduced that one twice.

**Comment by iamteapot422** [comment]
@stilnat sent answers in discord

**Comment by stilnat** [comment]
So the error breaking interaction seems to be present solely here, not on devellop, but it's very unreliable.
I just spammed the tilemap menu, picked up items and spawned them around 300 times, placed some turf, tile floors, locker, run around, spammed buttons, and that did not happen again. Very unreliable bug. 
That said, I did reproduce it 7 or 8 times, sometimes after 30 seconds of play time, so it's too much of an issue to accept it as it is. Relaunching Unity does not help.

**Review by stilnat — CHANGES_REQUESTED** [review]
nice, does those last commits fix the issue with the interaction system breaking ?

**Comment by iamteapot422** [comment]
@stilnat i think yes, couldn't reproduce it this time

**Comment by iamteapot422** [comment]
@stilnat added input consumption, check the pr's description

**Comment by stilnat** [comment]
before merging, must check if those bugs happen on develop as well, or if they stop happening on this one.
I believe the first one might happen on develop but can't reproduce it. The second only happen on this branch from what I saw.

BUG 1 : THE GHOST BUG

open the tilemap menu, spam some cans, spam right/left click on cans, pick up drop, repetitively until you can't pick up or drop with left click (but still work with right click). You get this exception. 

> **NullReferenceException: Object reference not set to an instance of an object
> SS3D.Systems.Tile.TileMapCreator.GhostManager.ChangeGhostColor (SS3D.Systems.Tile.TileMapCreator.GhostManager+BuildMatMode mode) (at Assets/Scripts/SS3D/Systems/Tile/TileMapCreator/GhostManager.cs:92)
> SS3D.Systems.Tile.UI.TileMapCreator.RefreshGhost () (at Assets/Scripts/SS3D/Systems/Tile/TileMapCreator/TileMapCreator.cs:311)
> SS3D.Systems.Tile.UI.TileMapCreator.HandleMouseClick (UnityEngine.Vector3 snappedPosition, System.Boolean replaceExisting) (at Assets/Scripts/SS3D/Systems/Tile/TileMapCreator/TileMapCreator.cs:181)
> SS3D.Systems.Tile.UI.TileMapCreator.HandlePlace (UnityEngine.InputSystem.InputAction+CallbackContext context) (at Assets/Scripts/SS3D/Systems/Tile/TileMapCreator/TileMapCreator.cs:110)
> UnityEngine.InputSystem.Utilities.DelegateHelpers.InvokeCallbacksSafe[TValue] (UnityEngine.InputSystem.Utilities.CallbackArray`1[System.Action`1[TValue]]& callbacks, TValue argument, System.String callbackName, System.Object context) (at Library/PackageCache/com.unity.inputsystem@1.4.4/InputSystem/Utilities/DelegateHelpers.cs:46)
> UnityEngine.InputSystem.LowLevel.<>c__DisplayClass7_0:<set_onUpdate>b__0(NativeInputUpdateType, NativeInputEventBuffer*)
> UnityEngineInternal.Input.NativeInputSystem:NotifyUpdate(NativeInputUpdateType, IntPtr)**

and this one 
> 
> NullReferenceException while executing 'performed' callbacks of 'Tile Creator/Place[/Mouse/leftButton]'
> UnityEngine.InputSystem.LowLevel.NativeInputRuntime/<>c__DisplayClass7_0:<set_onUpdate>b__0 (UnityEngineInternal.Input.NativeInputUpdateType,UnityEngineInternal.Input.NativeInputEventBuffer*)
> UnityEngineInternal.Input.NativeInputSystem:NotifyUpdate (UnityEngineInternal.Input.NativeInputUpdateType,intptr)

BUG 2 : THE LEFT+RIGHT CLICK INTERACTION BREAK

try spamming left and right click at the same time on an item, after a while you can't pick drop them with left click. It doesn't throw any exceptions in logs.

**Review by stilnat — APPROVED** [review]
I tried really hard to break this with no success, code is legit and it solves what it says it solves. Great work !

**Pull request merged**